### PR TITLE
fix content-length desync in TextIOPayload

### PR DESCRIPTION
**Pull request by arshsmith1**
## What do these changes do?

`TextIOPayload` reports its size through the inherited `IOBasePayload.size`, which returns `os.fstat().st_size`. A text stream is decoded and re-encoded on its way to the wire, so that on-disk byte count is not what actually gets written: universal-newline translation (`\r\n` to `\n`) and any gap between the file's encoding and the payload encoding both change the length. Serving an attacker-influenced text file with `web.Response(body=open(path))` then sends a `Content-Length` that disagrees with the body. A file of `b"a\r\nb\r\nc\r\n"` declares 9 bytes but writes 6, leaving a keep-alive connection three bytes short, and a latin-1 file re-encoded to utf-8 is cut to the too-small length, slicing a character in half.

The size genuinely can't be known without reading and encoding the whole stream, so `TextIOPayload.size` now returns `None`, which selects chunked/close framing instead of a wrong length. While confirming that, I found the unbounded write loop could spin forever for utf-16/utf-32 bodies, because `"".encode("utf-16")` is the two-byte BOM rather than empty, so an EOF read never looked like EOF. The read helpers now short-circuit an empty read to `b""`.

## Are there changes in behavior for the user?

Text file bodies are now sent with chunked transfer-encoding rather than a `Content-Length` that only matched for pure-ASCII, unix-newline content. `StringIOPayload` and binary file payloads are untouched and keep their exact `Content-Length`.

## Is it a substantial burden for the maintainers to support this?

No. It is a `size` override plus a one-line EOF guard in each read helper, all inside `TextIOPayload`. Two existing size tests asserted the on-disk value and are updated to the corrected behavior, and regression tests cover both the short-write and truncation cases.

## Related issue number

N/A

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [ ] Documentation reflects the changes - N/A, no public API change
- [x] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
- [x] Add a new news fragment into the `CHANGES/` folder

### CODE REVIEW

**Review by noqt — COMMENTED** [review]
Disclosure: I work on [NOQT's Lumi Trace](https://github.com/noqt/Lumi-Trace). I ran Lumi Trace 0.10.1 against this exact head (`862e7728`); it ranked `TextIOPayload._read_and_available_len` first, and I then reproduced the path directly.

The new empty-read guard fixes the UTF-16 EOF loop, but a payload larger than one read is still corrupted because each non-empty chunk is encoded with a fresh encoder. Each `chunk.encode("utf-16")` emits another BOM.

Minimal shape:

```python
text = "a" * (DEFAULT_CHUNK_SIZE + 1)
payload = TextIOPayload(io.StringIO(text), encoding="utf-16")
await payload.write(writer)
body = b"".join(writer.chunks)
```

On this head:

```json
{"body_matches_single_encode": false, "bom_count": 2, "chunk_count": 2, "decoded_extra_codepoints": 1, "embedded_bom_index": 262144}
```

So the framing fix sends the whole body, but decoding it produces an extra `U+FEFF` at the chunk boundary. The current UTF-16 test stays below `DEFAULT_CHUNK_SIZE`, so it doesn't exercise this.

I'd keep one incremental encoder for the entire write and flush it once at EOF (or explicitly reject stateful/BOM-emitting encodings), then add a test just over the chunk boundary. The `size = None` change itself still looks like the right framing direction.

File: aiohttp/payload.py [hunk 1]
```diff
@@ -724,6 +724,21 @@ class TextIOPayload(IOBasePayload):
     _value: io.TextIOBase
     # _autoclose = False (inherited) - Has text file handle that needs explicit closing
 
+    @property
+    def size(self) -> int | None:
```
Comment (Dreamsorcerer) [inline aiohttp/payload.py:728]: Don't place methods above init.

Comment (arshsmith1) [reply aiohttp/payload.py]: Done, moved it below __init__. On the per-chunk BOM note from the bots: that's pre-existing, each chunk was already encoded independently before this change, so I'd rather handle it with an incremental encoder in a follow-up unless you'd prefer it here.

File: aiohttp/payload.py [hunk 2]
```diff
@@ -751,6 +751,21 @@ def __init__(
...
+
+        Unlike the binary IOBasePayload, a text stream is decoded and
+        re-encoded before it reaches the wire, so os.fstat().st_size is not
+        the number of bytes actually written. Universal-newline translation
+        (``\\r\\n`` -> ``\\n``) and any difference between the on-disk encoding
+        and the payload encoding change the byte count, so the on-disk size
+        must not be advertised as Content-Length. Report the size as unknown
+        (which selects chunked/close framing) rather than a value that
+        disagrees with the body.
+        """
```
Comment (Dreamsorcerer) [inline aiohttp/payload.py]:
```suggestion
        """Size unknown as file bytes get re-encoded for a text stream."""
```

File: tests/test_payload.py [hunk 3]
```diff
@@ -1319,27 +1320,74 @@ async def test_text_io_payload_size_utf16(tmp_path: Path) -> None:
...
+
+
+async def test_text_io_payload_reencode_not_truncated(tmp_path: Path) -> None:
+    """Re-encoding to a wider form must not be truncated to the on-disk size.
+
+    A latin-1 source byte (``é`` -> 0xe9) becomes two bytes in utf-8. When the
+    on-disk size was used as Content-Length the body was cut to that length,
+    splitting the multibyte character; size must be None so the full utf-8
+    body is sent.
+    """
```
Comment (Dreamsorcerer) [inline tests/test_payload.py]:
```suggestion
    """Re-encoding to a wider form must not be truncated to the on-disk size."""
```

File: tests/test_payload.py [hunk 4]
```diff
@@ -1319,27 +1320,74 @@ async def test_text_io_payload_size_utf16(tmp_path: Path) -> None:
...
+    try:
+        tiop = payload.TextIOPayload(f)
+        assert tiop.size is None
+        writer = BufferWriter()
+        await tiop.write(writer)
+        assert bytes(writer.buffer) == b"a\nb\nc\n"
+    finally:
+        await loop.run_in_executor(None, f.close)
+
+
```
Comment (Dreamsorcerer) [inline tests/test_payload.py]:
```suggestion
```

File: aiohttp/payload.py [hunk 5]
```diff
@@ -784,6 +789,10 @@ def _read_and_available_len(
                 remaining_content_len or DEFAULT_CHUNK_SIZE,
             )
         )
+        # An empty read means EOF. Encoding "" is not always empty (utf-16/utf-32
+        # emit a BOM), so short-circuit to keep it a genuine end-of-stream marker.
+        if not chunk:
+            return size, b""
```
Comment (Dreamsorcerer) [inline aiohttp/payload.py:795]: Missing test coverage
